gadgets/traceloop/go/program.go

```diff
@@ -0,0 +1,649 @@
+	}
+
+	return nil
+}
+
+func fromCString(in []byte) string {
+	for i := 0; i < len(in); i++ {
+		if in[i] == 0 {
+			return string(in[:i])
+		}
+	}
+	return string(in)
+}
+
+func fromCStringN(in []byte, length int) string {
+	l := len(in)
+	if length < l {
+		l = length
+	}
+
+	for i := 0; i < l; i++ {
+		if in[i] == 0 {
+			return string(in[:i])
+		}
+	}
+	return string(in[:l])
+}
+
+func wallTimeFromBootTime(ts uint64) eventtypes.Time {
+	if ts == 0 {
+		return eventtypes.Time(time.Now().UnixNano())
+	}
+	return eventtypes.Time(time.Unix(0, int64(ts)).Add(0/*timeDiff*/).UnixNano())
+}
+
+func timestampFromEvent(event *syscallEvent) eventtypes.Time {
+	return wallTimeFromBootTime(event.bootTimestamp)
+}
+
+// Copied/pasted/adapted from kernel macro round_up:
+// https://elixir.bootlin.com/linux/v6.0/source/include/linux/math.h#L25
+func roundUp(x, y uintptr) uintptr {
+	return ((x - 1) | (y - 1)) + 1
+}
+
+// The kernel aligns size of perf event with the following snippet:
+// void perf_prepare_sample(...)
+//
+//	{
+//		//...
+//		size = round_up(sum + sizeof(u32), sizeof(u64));
+//		raw->size = size - sizeof(u32);
+//		frag->pad = raw->size - sum;
+//		// ...
+//	}
+//
+// (https://elixir.bootlin.com/linux/v6.0/source/kernel/events/core.c#L7353)
+// In the case of our structure of interest (i.e. struct_syscall_event_t and
+// struct_syscall_event_cont_t), their size will be increased by 4, here is
+// an example for struct_syscall_event_t which size is 88:
+// size = round_up(sum + sizeof(u32), sizeof(u64))
+//
+//	= round_up(88 + 4, 8)
+//	= round_up(92, 8)
+//	= 96
+//
+// raw->size = size - sizeof(u32)
+//
+//	= 96 - 4
+//	= 92
+//
+// So, 4 bytes will be added as padding at the end of the event and the size we
+// will read getting perfEventSample will be 92 instead of 88.
+func alignSize(structSize uintptr) uintptr {
+	var ret uintptr
+	var foo uint64
+	var bar uint32
+
+	ret = roundUp(structSize+unsafe.Sizeof(bar), unsafe.Sizeof(foo))
+	ret = ret - unsafe.Sizeof(bar)
+
+	return ret
+}
+
+// Convert a return value to corresponding error number if meaningful.
+// See man syscalls:
+// Note:
+// system calls indicate a failure by returning a negative error
+// number to the caller on architectures without a separate error
+// register/flag, as noted in syscall(2); when this happens, the
+// wrapper function negates the returned error number (to make it
+// positive), copies it to errno, and returns -1 to the caller of
+// the wrapper.
+func retToStr(ret uint64) string {
+	errNo := int64(ret)
+	if errNo >= -4095 && errNo <= -1 {
+		return fmt.Sprintf("-1 (%s)", syscall.Errno(-errNo).Error())
+	}
+	return fmt.Sprintf("%d", ret)
+}
+
+func (t *tracelooper) read(reader *containerRingReader) ([]*tracelooptypes.Event, error) {
+	syscallContinuedEventsMap := make(map[uint64][]*syscallEventContinued)
+	syscallEnterEventsMap := make(map[uint64][]*syscallEvent)
+	syscallExitEventsMap := make(map[uint64][]*syscallEvent)
```

Comment:
What I meant is, why to have `syscallEnterEventsMap := make(map[uint64][]*syscallEvent)` instead of `syscallEnterEventsMap := make(map[uint64]*syscallEvent)`? there should be a single enter event for each timestmap, right?